vcfindex.c

```diff
@@ -63,27 +63,37 @@ int vcf_index_stats(char *fname, int stats)
     const char **seq;
     int i, nseq;
     tbx_t *tbx = NULL;
+    bcf_hdr_t *hdr = NULL;
     hts_idx_t *idx = NULL;
 
     htsFile *fp = hts_open(fname,"r");
-    if ( !fp ) { fprintf(stderr,"Could not read %s\n", fname); return 1; }
-    bcf_hdr_t *hdr = bcf_hdr_read(fp);
-    if ( !hdr ) { fprintf(stderr,"Could not read the header: %s\n", fname); return 1; }
-
-    if ( hts_get_format(fp)->format==vcf )
-    {
-        tbx = tbx_index_load(fname);
-        if ( !tbx ) { fprintf(stderr,"Could not load index for VCF: %s\n", fname); return 1; }
-    }
-    else if ( hts_get_format(fp)->format==bcf )
-    {
-        idx = bcf_index_load(fname);
-        if ( !idx ) { fprintf(stderr,"Could not load index for BCF file: %s\n", fname); return 1; }
+    if ( !fp ) {
```

Comment:
The error message from `hts_open` feels messy. 

It looks like `hts_idx_load3` works irrespective of whether we give it a bcf, bcf.csi, vcf.gz, vcf.gz.csi or vcf.gz.tbi.  I didn't check .bai as it looks like it's no longer capable of being written by bcftools so testing it is rather challenging.

I think the old code is confusing you too.  It uses .tbi for vcf and .bai/.csi for bcf, but vcf can have .bai/.csi too and it works just fine.  Did the old code simply not support .csi indices for vcf?  Furthermore it looks like `idx_index_load3` can be given a .tbi file anyway, so is `tbx_index_load3` a legacy interface now?  Furthermore it doesn't winge if given a .bcf or .vcf.gz file, unlike `tbx_index_load3`.

I'm not sure I understand why we try both APIs.  Just historical reasons, or am I missing something?

Reply:
The error message from `hts_open` **is** messy: of all the API functions that can fail, opening files is the one that practically every calling program does actually check and print an error message for itself. IMHO `hts_open()` and the functions under it should not be printing messages themselves, but leave that to the caller — which knows the context.

Or even multiple spurious messy messages, e.g.,
```
$ samtools view htslib
[E::hts_hopen] Failed to open file htslib
[E::hts_open_format] Failed to open file "htslib" : Is a directory
samtools view: failed to open "htslib" for reading: Is a directory
```

Reply:
Ok what I was missing is the `--stats` output has chr length.  For some reason I had it in my head that this was in the undocumented CSI meta-data, but I can't find it with hd so I guess not.

However for `bcftools --nrecords` I can get the correct result for bcf, bcf.csi, vcf.gz, vcf.gz.csi and vcf.gz.tbi with the `htsFile *fp = hts_open(fname,"r");` line replaced by `htsFile *fp = NULL` (ie forced failure).  Furthermore commenting out the `tbx = tbx_index_load3...` line has no effect and it works just fine, so that's a red herring now it looks.

Sadly however in order get chromosome lengths from `--stats` we do have to open the main file up, which means we need to know whether to open it.  As there's no silent option (maybe something we should consider adding to htslib) nor is there a way of querying what type of index we loaded, so I think the only viable alternative for error-report free code therefore is checking file extensions.

Eg:

```
    char *d = strrchr(fn, '.');
    htsFile *fp = (d && (strcmp(d, ".tbi") == 0 ||
			 strcmp(d, ".bai") == 0 ||
			 strcmp(d, ".csi") == 0))
	? NULL
	: hts_open(fn, "r");
```

Assuming my code is correct, it'll not attempt to `hts_open` a file ending in .tbi, .bai or .csi, which avoids the bogus error message.

Reply:
> I'm not sure I understand why we try both APIs. Just historical reasons, or am I missing something?

I followed the logic of the existing code, but I will try to do away with `tbx_index_load`.

> IMHO hts_open() and the functions under it should not be printing messages themselves, but leave that to the caller — which knows the context.

This is an option worth considering. My secondary solution was to add a silent mode (`s`) to `hts_open`, which would make it not log errors.

Reply:
Argh it's even more confusing.  I'd forgotten what a total nightmare our "twisty maze of index formats all like" really is.

So `tbx_index_load3` can read .tbi and .vcf.gz.csi but not .bcf.csi.  From vague recollection that's because CSI is actually two different formats - one for VCF and one for BCF.  (The bit that differs is, of course, the undocumented aux part.)

With `hts_index_load3` instead, it works on some .tbi files but not on others.  I have no clue what the difference is.   Eg (with hts_index_load3 commented out and tbx_index_load3 being used always):

```
$ bcftools index --stats /nfs/users/nfs_j/jkb/scratch/data/ALL.chr20.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi
[E::hts_hopen] Failed to open file /nfs/users/nfs_j/jkb/scratch/data/ALL.chr20.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi
[E::hts_open_format] Failed to open file "/nfs/users/nfs_j/jkb/scratch/data/ALL.chr20.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.tbi" : Exec format error
20	.	1812841
```

With the vice-versa strategy, of tbx_index_load3 used and hts_index_load3 commented out we instead get:

```
$ bcftools index --stats /nfs/users/nfs_j/jkb/scratch/data/ALL.chr20.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.csi
[E::hts_hopen] Failed to open file /nfs/users/nfs_j/jkb/scratch/data/ALL.chr20.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.csi
[E::hts_open_format] Failed to open file "/nfs/users/nfs_j/jkb/scratch/data/ALL.chr20.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz.csi" : Exec format error
(null)	.	1812841
```

So it's lost the chromosome name.  Infact this happens to the .bcf file too.

Hence, sadly, I think we're going to have to check filenames and call the appropriate function if we don't want bogus error messages (eg as tbx_index_load3 gives when given the .bcf.csi file).

I'm starting to think the premise of this PR is wrong.  It's OK for `--nrecords`, but there's simply not enough data in the indices to work for `--stats`.

Reply:
Digging deeper, it is indeed CSI's schizophrenic nature which is causing that `(null)` for chromosome name.

vcf.gz.csi has the aux array as a tbi header block, which contains a mapping of IDs to names.  bcf.csi has a different format aux array that doesn't have this capability.

Practically speaking, the entire rest of the world uses .vcf.gz.tbi it seems as I believe we're the only package to have CSI indices?  So maybe the original requester wasn't aware of this.